File: 2015/01/19/List/Bowling.hs
module Bowling
where

score = score' 10 

score' 0    _ = 0
score' 1 [x,y] = x + y 
score' n (x:y:z:throws)
    | strike    = x + y + z + score' (n-1) (y:z:throws) 
    | spare     = x + y + z + score' (n-1) (z:throws)
    | otherwise = x + y     + score' (n-1) (z:throws)
    where 
        spare  = x+y == 10
        strike = x   == 10 

File: 2026/06/17/bowling.pl
% bowling.pl
:- module(bowling, [score/2]).

score([], 0).

% strike rule
score([Roll1 , Roll2 , Roll3 | Rest], Score) :-
    10 is Roll1,
    score([Roll2, Roll3 | Rest], ScoreRest),
    Score is 10 + Roll2 + Roll3 + ScoreRest.

% spare rule
score([Roll1 , Roll2 , Roll3 | Rest], Score) :-
    Roll1 \= 10,
    10 is Roll1 + Roll2,
    score([Roll3 | Rest], ScoreRest),
    Score is 10 + Roll3 + ScoreRest.

% average roll
score([Roll|Rest], Score) :-
    Roll \= 10,
    score(Rest, ScoreRest),
    Score is Roll + ScoreRest.

